src/orchestrator-kubernetes/src/lib.rs

```diff
@@ -448,11 +512,93 @@ impl NamespacedOrchestrator for NamespacedKubernetesOrchestrator {
+                            .await,
+                    )
+                } else {
+                    None
+                }
+            };
+            let disk_capacity_fut = async {
+                if disk {
+                    Some(
+                        self_
+                            .custom_metrics_api
+                            .get_subresource(
+                                "kubelet_volume_stats_capacity_bytes",
+                                &format!("{name}-scratch"),
+                            )
+                            .await,
+                    )
+                } else {
+                    None
+                }
+            };
+            let (metrics, disk_usage, disk_capacity) = match futures::future::join3(
+                self_.metrics_api.get(&name),
+                disk_usage_fut,
+                disk_capacity_fut,
+            )
+            .await
+            {
+                (Ok(metrics), disk_usage, disk_capacity) => {
+                    // TODO(guswynn): don't tolerate errors here, when we are more
+                    // comfortable with `prometheus-adapter` in production
+                    // (And we run it in ci).
+                    (
+                        metrics,
+                        disk_usage.and_then(|r| {
+                            r.map_err(|e| {
+                                match e {
+                                    Error::Api(e) if e.code == 404 => {
+                                        // The metric has not been populated yet.
+                                    }
+                                    _ => {
+                                        warn!(
+                                            "Failed to fetch \
+                                            `kubelet_volume_stats_used_bytes` for {name}: {e}"
+                                        )
+                                    }
+                                }
+                            })
+                            .ok()
+                        }),
+                        disk_capacity.and_then(|r| {
+                            r.map_err(|e| {
+                                match e {
+                                    Error::Api(e) if e.code == 404 => {
+                                        // This is either a replica without a disk, or the metric has
+                                        // been populated yet.
+                                    }
+                                    _ => {
+                                        warn!(
+                                            "Failed to fetch \
+                                            `kubelet_volume_stats_capacity_bytes` for {name}: {e}"
+                                        )
+                                    }
+                                }
+                            })
+                            .ok()
+                        }),
+                    )
```

Comment:
I find it really hard to reason about these sorts of nested combinators! Especially given that `map_err` is just used for its side effects. Can we do match statements here instead?

```suggestion
                    // TODO(guswynn): don't tolerate errors here, when we are more
                    // comfortable with `prometheus-adapter` in production
                    // (And we run it in ci).

                    let disk_usage = match disk_usage {
                        Some(Ok(disk_usage)) => Some(disk_usage),
                        Some(Err(e)) if !matches!(&e, Error::Api(e) if e.code == 404) => {
                            warn!("Failed to fetch `kubelet_volume_stats_used_bytes` for {name}: {e}");
                            None
                        }
                        _ => None,
                    };

                    let disk_capacity = match disk_capacity {
                        Some(Ok(disk_capacity)) => Some(disk_capacity),
                        Some(Err(e)) if !matches!(&e, Error::Api(e) if e.code == 404) => {
                            warn!("Failed to fetch `kubelet_volume_stats_capacity_bytes` for {name}: {e}");
                            None
                        }
                        _ => None,
                    };

                    (metrics, disk_usage, disk_capacity)
```

Reply:
yah this looks nice!